src/BotPlutusInterface/TimeSlot.hs

```diff
@@ -0,0 +1,203 @@
+{- | This module provides conversions between `Slot` and `POSIXTime`.
+ It uses node queries to get additional parameters and emulate conversions that happen in the ledger
+ when it builds `TxInfo`. It aims to provide reliable conversion, enabling execution of contracts
+ with exact time intervals validation, and should not be affected by slot length changes unlike
+ current plutus-ledger conversion functions.
+-}
+module BotPlutusInterface.TimeSlot (
+  TimeSlotConversionError,
+  slotToPOSIXTimeIO,
+  posixTimeToSlotIO,
+  posixTimeRangeToContainedSlotRangeIO,
+) where
+
+import BotPlutusInterface.QueryNode (
+  NodeInfo (NodeInfo),
+  queryEraHistory,
+  querySystemStart,
+ )
+import BotPlutusInterface.Types (
+  PABConfig,
+  pcNetwork,
+  pcProtocolParams,
+ )
+import Cardano.Api (CardanoMode, EraHistory)
+import Cardano.Api qualified as CAPI
+import Cardano.Ledger.Alonzo.PParams (_protocolVersion)
+import Cardano.Ledger.Alonzo.TxInfo (slotToPOSIXTime)
+import Cardano.Ledger.Slot (EpochInfo)
+import Cardano.Slotting.EpochInfo (hoistEpochInfo)
+import Cardano.Slotting.Time (SystemStart, toRelativeTime)
+import Control.Monad.Except (runExcept)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Trans.Either (
+  EitherT,
+  firstEitherT,
+  hoistEither,
+  newEitherT,
+  runEitherT,
+ )
+import Data.Bifunctor (first)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Time (UTCTime, secondsToNominalDiffTime)
+import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
+import Ledger (
+  Extended (Finite, NegInf, PosInf),
+  Interval (Interval),
+  LowerBound (LowerBound),
+  UpperBound (UpperBound),
+ )
+import Ledger qualified
+import Ouroboros.Consensus.HardFork.History qualified as Consensus
+import Ouroboros.Consensus.HardFork.History.Qry qualified as HF
+import System.Environment (getEnv)
+import Prelude
+
+-- | Error returned by the functions of this module
+data TimeSlotConversionError
+  = TimeSlotConversionError !Text
+  deriving stock (Show)
+
+-- Analogous to `slotToBeginPOSIXTime` from plutus-ledger
+
+-- | Convert `Slot` to `POSIXTime`.
+slotToPOSIXTimeIO :: PABConfig -> Ledger.Slot -> IO (Either TimeSlotConversionError Ledger.POSIXTime)
+slotToPOSIXTimeIO pabConf lSlot = runEitherT $ do
+  nodeInfo <- liftIO $ mkNodeInfo pabConf
+  eraHsitory <- newET (queryEraHistory nodeInfo)
```

Comment:
I thought Sam mentioned this, but i guess that was some other place.
```suggestion
  eraHistory <- newET (queryEraHistory nodeInfo)
```

Reply:
fixed